### ISSUE TITLE
 Misterious Panic in Snake Game Example

### ISSUE THREAD

**Issue by brunoczim** [brunoczim/andiskaz#1]
When you execute the example snakegame:
```sh
cargo run --example=snakegame
```
If you attempt to reduce the size of the screen, it is possible that a panic happens. Why? I don't know exactly why, but I have a clue: perhaps it is because a resize event gets shadowed by a key event. I am saying this cause I guess every time this happened I was pressing a key, but I am not sure.

It is highly possible this is indeed the case, since only the last event is available in the events channel, and that if the game doesn't account for the actual screen size, it might write outside the screen bounds, which is the case. This is how the panic looks like:
```
cargo run --example=snakegame
    Finished dev [unoptimized + debuginfo] target(s) in 0.04s
     Running `target/debug/examples/snakegame`

panicked at 'Screen out of bounds', src/screen/locked.rs:74:43
   0: snakegame::main::{{closure}}::{{closure}}
             at examples/snakegame/main.rs:35:25
   1: std::panicking::rust_panic_with_hook
             at /rustc/b1af43bc63bc7417938df056f7f25d456cc11b0e/library/std/src/panicking.rs:581:17
   2: std::panicking::begin_panic_handler::{{closure}}
             at /rustc/b1af43bc63bc7417938df056f7f25d456cc11b0e/library/std/src/panicking.rs:484:9
   3: std::sys_common::backtrace::__rust_end_short_backtrace
             at /rustc/b1af43bc63bc7417938df056f7f25d456cc11b0e/library/std/src/sys_common/backtrace.rs:153:18
   4: rust_begin_unwind
             at /rustc/b1af43bc63bc7417938df056f7f25d456cc11b0e/library/std/src/panicking.rs:483:5
   5: core::panicking::panic_fmt
             at /rustc/b1af43bc63bc7417938df056f7f25d456cc11b0e/library/core/src/panicking.rs:85:14
   6: core::option::expect_failed
             at /rustc/b1af43bc63bc7417938df056f7f25d456cc11b0e/library/core/src/option.rs:1229:5
   7: core::option::Option<T>::expect
             at /home/bruno/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/option.rs:349:21
   8: andiskaz::screen::locked::LockedScreen::update
             at src/screen/locked.rs:74:13
   9: andiskaz::screen::locked::LockedScreen::set
             at src/screen/locked.rs:53:9
  10: snakegame::snake::Snake::render
             at examples/snakegame/snake.rs:113:13
  11: snakegame::game::Game::render
             at examples/snakegame/game.rs:290:9
  12: snakegame::game::Game::run::{{closure}}
             at examples/snakegame/game.rs:154:13
  13: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /home/bruno/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/future/mod.rs:80:19
  14: snakegame::term_main::{{closure}}
             at examples/snakegame/main.rs:64:20
  15: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /home/bruno/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/future/mod.rs:80:19
  16: andiskaz::terminal::main_task::{{closure}}
             at src/terminal.rs:208:5
  17: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /home/bruno/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/future/mod.rs:80:19
  18: tokio::runtime::task::core::Core<T,S>::poll::{{closure}}
             at /home/bruno/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.3.2/src/runtime/task/core.rs:173:17
  19: tokio::loom::std::unsafe_cell::UnsafeCell<T>::with_mut
             at /home/bruno/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.3.2/src/loom/std/unsafe_cell.rs:14:9
  20: tokio::runtime::task::core::Core<T,S>::poll
             at /home/bruno/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.3.2/src/runtime/task/core.rs:158:13
  21: tokio::runtime::task::harness::Harness<T,S>::poll::{{closure}}
             at /home/bruno/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.3.2/src/runtime/task/harness.rs:107:27
  22: core::ops::function::FnOnce::call_once
             at /home/bruno/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/ops/function.rs:227:5
  23: <std::panic::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once
             at /home/bruno/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/panic.rs:308:9
  24: std::panicking::try::do_call
             at /home/bruno/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/panicking.rs:381:40
  25: __rust_try
  26: std::panicking::try
             at /home/bruno/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/panicking.rs:345:19
  27: std::panic::catch_unwind
             at /home/bruno/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/panic.rs:382:14
  28: tokio::runtime::task::harness::Harness<T,S>::poll
             at /home/bruno/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.3.2/src/runtime/task/harness.rs:89:19
  29: tokio::runtime::task::raw::poll
             at /home/bruno/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.3.2/src/runtime/task/raw.rs:104:5
  30: tokio::runtime::task::raw::RawTask::poll
             at /home/bruno/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.3.2/src/runtime/task/raw.rs:66:18
  31: tokio::runtime::task::Notified<S>::run
             at /home/bruno/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.3.2/src/runtime/task/mod.rs:171:9
  32: tokio::runtime::thread_pool::worker::Context::run_task::{{closure}}
             at /home/bruno/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.3.2/src/runtime/thread_pool/worker.rs:346:13
  33: tokio::coop::with_budget::{{closure}}
             at /home/bruno/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.3.2/src/coop.rs:121:9
  34: std::thread::local::LocalKey<T>::try_with
             at /home/bruno/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/thread/local.rs:272:16
  35: std::thread::local::LocalKey<T>::with
             at /home/bruno/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/thread/local.rs:248:9
  36: tokio::coop::with_budget
             at /home/bruno/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.3.2/src/coop.rs:114:5
      tokio::coop::budget
             at /home/bruno/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.3.2/src/coop.rs:98:5
      tokio::runtime::thread_pool::worker::Context::run_task
             at /home/bruno/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.3.2/src/runtime/thread_pool/worker.rs:345:9
  37: tokio::runtime::thread_pool::worker::Context::run
             at /home/bruno/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.3.2/src/runtime/thread_pool/worker.rs:317:24
  38: tokio::runtime::thread_pool::worker::run::{{closure}}
             at /home/bruno/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.3.2/src/runtime/thread_pool/worker.rs:302:17
  39: tokio::macros::scoped_tls::ScopedKey<T>::set
             at /home/bruno/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.3.2/src/macros/scoped_tls.rs:61:9
  40: tokio::runtime::thread_pool::worker::run
             at /home/bruno/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.3.2/src/runtime/thread_pool/worker.rs:299:5
  41: tokio::runtime::thread_pool::worker::block_in_place::{{closure}}::{{closure}}
             at /home/bruno/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.3.2/src/runtime/thread_pool/worker.rs:255:41
  42: <tokio::runtime::blocking::task::BlockingTask<T> as core::future::future::Future>::poll
             at /home/bruno/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.3.2/src/runtime/blocking/task.rs:41:21
  43: tokio::runtime::task::core::Core<T,S>::poll::{{closure}}
             at /home/bruno/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.3.2/src/runtime/task/core.rs:173:17
  44: tokio::loom::std::unsafe_cell::UnsafeCell<T>::with_mut
             at /home/bruno/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.3.2/src/loom/std/unsafe_cell.rs:14:9
  45: tokio::runtime::task::core::Core<T,S>::poll
             at /home/bruno/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.3.2/src/runtime/task/core.rs:158:13
  46: tokio::runtime::task::harness::Harness<T,S>::poll::{{closure}}
             at /home/bruno/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.3.2/src/runtime/task/harness.rs:107:27
  47: core::ops::function::FnOnce::call_once
             at /home/bruno/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/ops/function.rs:227:5
  48: <std::panic::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once
             at /home/bruno/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/panic.rs:308:9
  49: std::panicking::try::do_call
             at /home/bruno/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/panicking.rs:381:40
  50: __rust_try
  51: std::panicking::try
             at /home/bruno/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/panicking.rs:345:19
  52: std::panic::catch_unwind
             at /home/bruno/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/panic.rs:382:14
  53: tokio::runtime::task::harness::Harness<T,S>::poll
             at /home/bruno/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.3.2/src/runtime/task/harness.rs:89:19
  54: tokio::runtime::task::raw::poll
             at /home/bruno/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.3.2/src/runtime/task/raw.rs:104:5
  55: tokio::runtime::task::raw::RawTask::poll
             at /home/bruno/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.3.2/src/runtime/task/raw.rs:66:18
  56: tokio::runtime::task::Notified<S>::run
             at /home/bruno/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.3.2/src/runtime/task/mod.rs:171:9
  57: tokio::runtime::blocking::pool::Inner::run
             at /home/bruno/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.3.2/src/runtime/blocking/pool.rs:252:17
  58: tokio::runtime::blocking::pool::Spawner::spawn_thread::{{closure}}
             at /home/bruno/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.3.2/src/runtime/blocking/pool.rs:233:17
```

**Comment by brunoczim** [comment]
It seems like a key is not necessarily needed.

**Comment by brunoczim** [comment]
It seems like the game is not handling the snake properly when resizing

**Comment by brunoczim** [comment]
It seems to be caused in first place indeed because of snake out of bounds. The problem is that the snake was being tested as being out of bounds, nothing was done about it and later the game would try to render the snake. I fixed it by saturating snake points when checking if the snake was out of bounds.

However, I later found out the Food suffered from the same issue. The problem is that if snake was out of bounds, the function would return early, and the food would not be regenerated in a valid position. I fixed this by *not* early returning.
